ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,412 @@
+
+
+## Errors
+
+Here's the Error hierarchy that is planned for the Ruby SDK:
+
+```ruby
+module Temporal
+  # top level error superclass
+  class Error < StandardError
+
+  # superclass for all errors within the SDK itself
+  class InternalError < Error
+
+  # errors specific to Client
+  class ClientError < Error
+
+  # errors specific to Worker
+  class WorkerError < InternalError
+
+  # superclass for connection, network and response errors
+  class RPCError < Error
+  class UnexpectedError < RPCError
+
+  # superclass for all API failure responses
+  class APIError < ServerError
+
+  # type hinting errors
+  class TypeError < Error
+
+  # superclass for workflow errors
+  class WorkflowError < Error
+```
+
+Notes:
+
+- This list is not exhaustive, more errors will be added as parts of the SDK are getting built
+- Some concrete error class placement might be confusing and will be clarified later in the process
+
+
+### Converting API Failures
+
+Similar to other SDKs we will decode `Failure` proto messages to native error classes. All `cause`s
+will be decoded recursively and the result will form a hierarchy identical to the one in the proto
+response. These error classes will all be subclasses of an `APIError` and will have names matching
+the original protobuf messages (while replacing `Failure` with `Error` for consistency within Ruby):
+
+```ruby
+module Temporal
+  class ApplicationError < APIError
+  class TimeoutError < APIError
+  class CanceledError < APIError
+  class TerminatedError < APIError
+  class ServerError < APIError
+  class ResetWorkflowError < APIError
+  class ActivityError < APIError
+  class ChildWorkflowExecutionError < APIError
+end
+```
+
+At this point we are not going to be converting user errors back to their original classes. These
+will be decoded into `ApplicationError`s while keeping the original class name, message, stack trace
+and other details. Here's an example of handling an activity error:
+
+```ruby
+class ProcessPaymentActivity < Temporal::Activity
+  class InsufficientFunds < Temporal::NonRetryableError; end
+  class PaymentNetworkError < Temporal::RetryableError
+    attr_reader :status
+
+    def initialize(message, status)
+      super(message)
+      @status = status
+    end
+  end
+
+  def execute(account, amount)
+    if amount < balance_of(account.number)
+      raise InsufficientFunds, "Account #{account.number} does not have enough funds"
+    end
+
+    response = process_payment(account.number, amount)
+    if response.failure?
+      raise PaymentNetworkError, response.message, response.status
+    end
+
+    ...
+  end
+end
+
+class RenewSubscriptionWorkflow < Temporal::Workflow
+  def execute
+    ...
+
+    begin
+      workflow.execute_activity(ProcessPaymentActivity, account, amount)
+    rescue Temporal::ActivityError => error
+      if error.cause.is_a?(Temporal::ApplicationError)
+        case error.cause.original_class
+        when 'ProcessPaymentActivity::InsufficientFunds'
```

Comment:
How do you know how many levels deep to qualify this? If this was `class MyError` in `MyCompany::MyTemporalWorkflows` module, what would this value be? In other SDKs, we have left this unqualified when storing on the `type` in `ApplicationError`.

Reply:
This will be the result of a `.class.name` call on the error itself. In Ruby it's quite clear, because you can't reference a class without all of the modules it's nested in:

```ruby
module MyCompany
  module MyTemporalWorkflows
    class MyError < StandardError; end

    puts MyError.name # prints "MyCompany::MyTemporalWorkflows::MyError"
  end
end
```

Reply:
Java also does this which is qualified, but others may not. This has never been a problem before because we have never used "type" to reference a class. This would be the first time "type" is ever used that way that I know of which is why I think we need a more general discussion. "type" is just a general user string, not meant for machine reading or having any expectations around it. Often "type" is just used in places like retry policy's non-retryable-types.

Reply:
> This will be the result of a .class.name call on the error itself.

:+1:

> you can't reference a class

We have never used "type" to reference a class and if we're gonna start doing that, we need a more general discussion. "type" is a user string that has no expected format and is not meant for machine reading. It just happens to default to the class name, but it's really meant to be an "error type", often used for matching with user-defined strings in retry policy's non-retryable types.

Reply:
What would be the expected way for the user to set the `type` field if it's a user string? Are they expected to always `raise ApplicationError.new(type: 'my-error-type')`?

Reply:
Yes. But our default failure converter defaults it to the error class name for them if they don't create an application error but we are still wrapping it as one (e.g. for activity failures).